client/src/main/scala/org/apache/celeborn/client/commit/CommitHandler.scala

```diff
@@ -180,6 +192,140 @@ abstract class CommitHandler(
+    var timeout = conf.rpcAskTimeout.duration.toMillis * maxRetries
+    val delta = 50
+    while (timeout >= 0 && !futures.isEmpty) {
+      val iter = futures.iterator()
+      while (iter.hasNext) {
+        val status = iter.next()
+        if (status.future.isCompleted) {
+          status.future.value.get match {
+            case scala.util.Success(res) =>
+              val worker = status.commitFilesParam.worker
+              res.status match {
+                case StatusCode.SUCCESS => // do nothing
+                case StatusCode.PARTIAL_SUCCESS | StatusCode.SHUFFLE_NOT_REGISTERED | StatusCode.REQUEST_FAILED | StatusCode.WORKER_EXCLUDED =>
+                  logInfo(s"Request commitFiles return ${res.status} for " +
+                    s"${Utils.makeShuffleKey(appUniqueId, shuffleId)}")
+                  if (res.status != StatusCode.WORKER_EXCLUDED) {
+                    commitFilesFailedWorkers.put(worker, (res.status, System.currentTimeMillis()))
+                  }
+                case _ =>
+                  logError(s"Should never reach here! commit files response status ${res.status}")
+              }
+
+              processResponse(res, worker)
+              iter.remove()
+            case scala.util.Failure(e) =>
+              val worker = status.commitFilesParam.worker
+              logError(
+                s"AskSync worker($worker) CommitFiles for $shuffleId failed" +
+                  s" (attempt ${status.retriedTimes}/$maxRetries).",
+                e)
+              if (status.retriedTimes < maxRetries) {
+                status.retriedTimes = status.retriedTimes + 1
+                status.future = commitFiles(
+                  appUniqueId,
+                  shuffleId,
+                  status.commitFilesParam.worker,
+                  status.commitFilesParam.primaryIds,
+                  status.commitFilesParam.replicaIds)
+              } else {
+                val res = CommitFilesResponse(
+                  StatusCode.REQUEST_FAILED,
+                  List.empty.asJava,
+                  List.empty.asJava,
+                  status.commitFilesParam.primaryIds,
+                  status.commitFilesParam.replicaIds)
+                processResponse(res, status.commitFilesParam.worker)
+                iter.remove()
+              }
+          }
+        }
+      }
+
+      if (!futures.isEmpty) {
+        Thread.sleep(delta)
+      }
+      timeout = timeout - delta
+    }
+
+    val iter = futures.iterator()
+    while (iter.hasNext) {
+      val status = iter.next()
+      logError(
+        s"AskSync worker(${status.commitFilesParam.worker}) CommitFiles for $shuffleId timed out")
+      val res = CommitFilesResponse(
+        StatusCode.REQUEST_FAILED,
+        List.empty.asJava,
+        List.empty.asJava,
+        status.commitFilesParam.primaryIds,
+        status.commitFilesParam.replicaIds)
+      processResponse(res, status.commitFilesParam.worker)
+      iter.remove()
```

Comment:
Scala's Future does not have cancel interface. I think it's OK not to call cancel, in this case the client can't cancel the future if it is already sent to worker.

Reply:
> it's OK not to call cancel

will that cause unexpected threads occupation and run out of the pool?

Reply:
> > it's OK not to call cancel
> 
> will that cause unexpected threads occupation and run out of the pool?

IMO cancelling a future means to interrupt the thread executing it. Since the operation is relatively lightweight, and because scala's Future does not offer interface to cancel, I think it's OK not to try to cancel it. Not to say previously askSync will not cancel the future. IIUC, calling `timeout.awaitResult(future)` only waits for some time and throws Exception if result is not ready, it will not cancel the future.

Reply:
> In the implementation of NettyRpcEnv, the ask method is using the thread that invoked the ask method to process the future's callback.

As discussed offline with @FMX , this is not the case. cc @pan3793

Reply:
> > In the implementation of NettyRpcEnv, the ask method is using the thread that invoked the ask method to process the future's callback.
> 
> As discussed offline with @FMX , this is not the case. cc @pan3793

thanks for clarifying.